### ISSUE TITLE
[HxGrid] Support for OnRowClick event

### ISSUE THREAD

**Issue by Arnnes** [havit/Havit.Blazor#552]
Would it be possible to add support for the OnRowClick event in HxGrid? The SelectedDataItemChanged event alone is not sufficient for my scenario. I need the ability to show the offcanvas when the user clicks on a row for the first time and hide it when the user clicks on the same row for the second time.

**Comment by hakenr** [comment]
Currently, the `role="button"` attribute renders to table-row (`<tr>`) only if there is `SelectionEnabled || MultiSelectionEnabled` (so does the event-handler).

Regarding the new `OnRowClick`:
1. Do you want the new callback to be available only for `SelectionEnabled || MultiSelectionEnabled`? _I guess no._
2. Do you want the clickable row to have `pointer` cursor? Currently brought by the `role="button"` attribute. _I guess yes._
3. Do you want the whole click-ability to be configurable, i.e. to have the `pointer` cursor set and `OnRowClick` invoked only when there is something like `RowClickable="true"`? _Probably yes._

Note for myself: Invoking `SelectedDataItemChanged` for cases when the `SelectedDataItem` is "changed" to the same item might also work as it would allow handling the callback like this:
```csharp
private async Task HandleSelectedDataItemChanged(TItem newSelectedItem)
{
  if (this.selectedItem == newSelectedItem)
  {
      this.selectedItem = null;
      await myOffcanvas.HideAsync();
  }
  else
  {
      this.selectedItem = newSelectedItem;
      ...
  }
}
```

💡 ...which shows me to a possible solution (breaking change?!): **What if we just adjust the HxGrid to "deselect" the row on second click (when it is already selected)?** Or maybe configurable with new `DeselectSelectedRowOnClick="true"` parameter? (Probably not needed.)

@crdo @Arnnes @jirikanda Your thoughts?

**Comment by Arnnes** [comment]
1. No, we need the new callback to be available even if `SelectionEnabled || MultiSelectionEnabled` is not set.
2. Yes, a pointer cursor would be nice.
3. I can see the usefulness of configuring the click-ability.

The implementation where `SelectedDataItemChanged` is invoked even when the same item is selected is sufficient. 

Modifying the `HxGrid` to automatically deselect the row on the second click could be a more elegant solution. However I think the configuration with new parameter `DeselectSelectedRowOnClick` might be necessary in certain scenarios.

**Comment by bholbrook** [comment]
I'd like to express my interest in this functionality too.
For me, I have scenarios where I want to hide/show something when the same row is clicked but because the SelectedDataItemChanged event isn't seemingly called when the selected item is clicked again this isn't possible at the moment.
For my scenario just having a simple boolean to change the behaviour to trigger a callback when the same item is selected would be sufficient.

**Comment by hakenr** [comment]
#923 now lets you use the `ItemRowAdditionalAttributesSelector` to add attributes to the `<tr>` (including event callbacks). 

Be careful not to conflict with attributes used by the `HxGrid` implementation (like `onclick`), but you can use `onmouseup` or similar to achieve your goal.

**Comment by gdube-nrc** [comment]
I just want to reference an example here, since it wasn't obvious to me how you would define an event in a dictionary.
[Example](https://github.com/havit/Havit.Blazor/blob/40fa00a80d56a3af9a81532b84e1751eb87d97c9/Havit.Blazor.TestApp/Havit.Blazor.TestApp.Client/HxGridTests/HxGrid_DragDropRows_Test.razor#L48):
```
Dictionary<string, object> EmployeeRowAttributes(Person item)
{
	return new() {
		{"ondragstart",  EventCallback.Factory.Create<DragEventArgs>(this, (e) => HandleDragStart(item, e))},
		{"ondrop", EventCallback.Factory.Create<DragEventArgs>(this, () => HandleDrop(item))},
		// {"ondragenter",   EventCallback.Factory.Create<DragEventArgs>(this,(e) => HandleDragEnter(item, e)) },
		// {"ondragleave",   EventCallback.Factory.Create<DragEventArgs>(this,HandleDragLeave) },
		// {"ondragend",  EventCallback.Factory.Create<DragEventArgs>(this,HandleDragEnd) },
	   };
}
```

**Comment by gdube-nrc** [comment]
I noticed that when adding onmouseup, when clicking on a HxButton within a row & using `OnClickStopPropagation`, it will still fire the row event. Which is obvious in hindsight, but it makes the workaround a bit less appealing. You'd have to disable HxGrid's selection feature and reimplement it yourself using RowAttributes with onclick.

**Comment by hakenr** [comment]
@gdube-nrc The `OnClickStopPropagation` parameter only affects the `onclick` event, not the `onmouseup` event (which still occurs on the button and propagates). What's the scenario you're trying to achieve?

**Comment by gdube-nrc** [comment]
Say it is a list of parts. I wanted to make the row clickable anywhere to view the part's details, but also have a button within the row to edit the part. Using the workaround, a click on the button will trigger both events. 

I remembered that you could turn off selection so that HxGrid doesn't use onclick itself, and then I could use it in RowAttributes. That works. I wonder how the API could be changed to make this a bit better, on the api consumer side.

**Comment by hakenr** [comment]
> I wanted to make the row clickable anywhere to view the part's details

You don't need `onmouseup` for this, just use the regular `SelectionEnabled="true"` (it is enabled by default, just don't disable it) and `@bind-SelectedDataItem="_selectedItem"` with `@bind-SelectedDataItem:after="HandleItemSelected"` to open the detail.

**Comment by gdube-nrc** [comment]
Yes, but then we circle back to the issue that we can't re-click on it to fire the event once it is selected.
And of course it feels like a hack.

**Comment by hakenr** [comment]
@gdube-nrc Sorry, I'm like an LLM, I lost the initial context :-D

Can you de-select the row in `@bind-SelectedDataItem:after="HandleItemSelected"` or do you need to keep the selection (keep the row highlighted)? Maybe you could use a custom `ItemRowCssClassSelector` to highlight the item with details displayed, and use `@bind-SelectedDataItem:after="HandleItemSelected"` just to take over the selection and reset it to `null`?

**Comment by hakenr** [comment]
[[Blazor] Add OnRowClick EventCallback to QuickGrid (#64602) · dotnet/aspnetcore@6b8e991](https://github.com/dotnet/aspnetcore/commit/6b8e99109a3a3f47c534a3c2b4233196e4b6cba5)

Blazor's `QuickGrid` recently added the `OnRowClick` event. We should probably add it too

* to allow easy transitions,
* to avoid confusing developers who might expect this event,
* to meet future expectations from AI tools.
